--- a/crypto/rand/getentropy.cc
+++ b/crypto/rand/getentropy.cc
@@ -26,6 +26,9 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <unistd.h>
+#if defined(__QNX__)
+#include <fcntl.h>
+#endif
 
 #if defined(OPENSSL_MACOS) || defined(OPENSSL_FUCHSIA)
 #include <sys/random.h>
@@ -35,6 +38,8 @@
 
 void bssl::CRYPTO_init_sysrand() {}
 
+#if !defined(__QNX__)
+
 // CRYPTO_sysrand puts |requested| random bytes into |out|.
 void bssl::CRYPTO_sysrand(uint8_t *out, size_t requested) {
   while (requested > 0) {
@@ -50,4 +55,26 @@
   }
 }
 
+#else
+
+void bssl::CRYPTO_sysrand(uint8_t *out, size_t requested) {
+  int fd = open("/dev/urandom", O_RDONLY);
+  if (fd < 0) {
+    abort();
+  }
+
+  while (requested > 0) {
+    ssize_t r = read(fd, out, requested);
+    if (r <= 0) {
+      abort();
+    }
+    out += r;
+    requested -= r;
+  }
+
+  close(fd);
+}
+
+#endif
+
 #endif  // OPENSSL_RAND_GETENTROPY
