--- Fast-DDS-2.14.4/test/unittest/transport/CMakeLists.txt	2026-03-25 13:01:45.000000000 +0000
+++ Fast-DDS-2.14.4_back/test/unittest/transport/CMakeLists.txt	2026-03-25 12:46:49.000000000 +0000
@@ -426,46 +426,48 @@
 ##########################
 # SHM tests
 ##########################
-if(IS_THIRDPARTY_BOOST_OK)
-    add_executable(SharedMemTests ${SHAREDMEMTESTS_SOURCE})
 
-    target_compile_definitions(SharedMemTests PRIVATE
-        BOOST_ASIO_STANDALONE
-        ASIO_STANDALONE
-        $<$<BOOL:${WIN32}>:_ENABLE_ATOMIC_ALIGNMENT_FIX>
-        $<$<AND:$<NOT:$<BOOL:${WIN32}>>,$<STREQUAL:"${CMAKE_BUILD_TYPE}","Debug">>:__DEBUG>
-        $<$<BOOL:${INTERNAL_DEBUG}>:__INTERNALDEBUG> # Internal debug activated.
-        )
+#if(IS_THIRDPARTY_BOOST_OK)
+#    add_executable(SharedMemTests ${SHAREDMEMTESTS_SOURCE})
+#
+#    target_compile_definitions(SharedMemTests PRIVATE
+#        BOOST_ASIO_STANDALONE
+#        ASIO_STANDALONE
+#        $<$<BOOL:${WIN32}>:_ENABLE_ATOMIC_ALIGNMENT_FIX>
+#        $<$<AND:$<NOT:$<BOOL:${WIN32}>>,$<STREQUAL:"${CMAKE_BUILD_TYPE}","Debug">>:__DEBUG>
+#        $<$<BOOL:${INTERNAL_DEBUG}>:__INTERNALDEBUG> # Internal debug activated.
+#        )
+#
+#    target_include_directories(SharedMemTests PRIVATE
+#        ${Asio_INCLUDE_DIR}
+#        ${PROJECT_SOURCE_DIR}/test/mock/rtps/MessageReceiver
+#        ${PROJECT_SOURCE_DIR}/test/mock/rtps/ReceiverResource
+#        ${PROJECT_SOURCE_DIR}/include ${PROJECT_BINARY_DIR}/include
+#        ${PROJECT_SOURCE_DIR}/src/cpp
+#        ${THIRDPARTY_BOOST_INCLUDE_DIR}
+#        $<$<BOOL:${ANDROID}>:${ANDROID_IFADDRS_INCLUDE_DIR}>
+#        )
+#    target_link_libraries(SharedMemTests
+#        foonathan_memory
+#        fastcdr
+#        GTest::gtest
+#        ${MOCKS}
+#        $<$<BOOL:${TLS_FOUND}>:OpenSSL::SSL$<SEMICOLON>OpenSSL::Crypto>
+#        ${THIRDPARTY_BOOST_LINK_LIBS}
+#        eProsima_atomic
+#        )
+#    if(QNX)
+#        target_link_libraries(SharedMemTests socket)
+#    endif()
+#    if(MSVC OR MSVC_IDE)
+#        target_link_libraries(SharedMemTests ${PRIVACY} iphlpapi Shlwapi )
+#    else()
+#        target_link_libraries(SharedMemTests ${PRIVACY} )
+#    endif()
+#    gtest_discover_tests(SharedMemTests)
+#endif()
+#
 
-    target_include_directories(SharedMemTests PRIVATE
-        ${Asio_INCLUDE_DIR}
-        ${PROJECT_SOURCE_DIR}/test/mock/rtps/MessageReceiver
-        ${PROJECT_SOURCE_DIR}/test/mock/rtps/ReceiverResource
-        ${PROJECT_SOURCE_DIR}/include ${PROJECT_BINARY_DIR}/include
-        ${PROJECT_SOURCE_DIR}/src/cpp
-        ${THIRDPARTY_BOOST_INCLUDE_DIR}
-        $<$<BOOL:${ANDROID}>:${ANDROID_IFADDRS_INCLUDE_DIR}>
-        )
-    target_link_libraries(SharedMemTests
-        foonathan_memory
-        fastcdr
-        GTest::gtest
-        ${MOCKS}
-        $<$<BOOL:${TLS_FOUND}>:OpenSSL::SSL$<SEMICOLON>OpenSSL::Crypto>
-        ${THIRDPARTY_BOOST_LINK_LIBS}
-        eProsima_atomic
-        )
-    if(QNX)
-        target_link_libraries(SharedMemTests socket)
-    endif()
-    if(MSVC OR MSVC_IDE)
-        target_link_libraries(SharedMemTests ${PRIVACY} iphlpapi Shlwapi )
-    else()
-        target_link_libraries(SharedMemTests ${PRIVACY} )
-    endif()
-    gtest_discover_tests(SharedMemTests)
-endif()
-
 # Add 'xfail' label to flaky tests
 set_property(DIRECTORY APPEND PROPERTY
     TEST_INCLUDE_FILES ${CMAKE_CURRENT_LIST_DIR}/xfail_tests.cmake
