--- arrow-apache_arrow_test_orig/cpp/src/arrow/CMakeLists.txt
+++ arrow-apache_arrow_test/cpp/src/arrow/CMakeLists.txt
@@ -713,7 +713,10 @@
   list(APPEND ARROW_TESTING_STATIC_LINK_LIBS "ws2_32")
   list(APPEND ARROW_TESTING_STATIC_INSTALL_INTERFACE_LIBS "ws2_32")
 endif()
-
+if(CMAKE_SYSTEM_NAME STREQUAL "QNX")
+  list(APPEND ARROW_TESTING_SHARED_LINK_LIBS socket)
+  list(APPEND ARROW_TESTING_STATIC_LINK_LIBS socket)
+endif()
 set(ARROW_TESTING_SRCS
     io/test_common.cc
     ipc/test_common.cc
