see https://github.com/qnx-ports/aports/issues/479

diff --git a/colcon_cmake/task/cmake/build.py b/colcon_cmake/task/cmake/build.py
index ad72839..2d5d346 100644
--- a/colcon_cmake/task/cmake/build.py
+++ b/colcon_cmake/task/cmake/build.py
@@ -309,6 +309,10 @@ class CmakeBuildTask(TaskExtensionPoint):
         if jobs is None:
             # the number of cores can't be determined
             return []
+
+        if not 'getloadavg' in os.__dict__:
+            return [f'-j{jobs}']
+
         return [
             f'-j{jobs}',
             f'-l{jobs}',
