diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0b9209c..aa4666e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -216,8 +216,8 @@ if(WITH_DLT_USE_IPv6)
     add_definitions(-DDLT_USE_IPv6)
 endif()
 
-if(WITH_DLT_QNX_SYSTEM AND NOT "${CMAKE_C_COMPILER}" MATCHES "nto-qnx|qcc|ntoaarch64-gcc|ntox86_64-gcc")
-    message(FATAL_ERROR "Can only compile for QNX with a QNX compiler, but found '${CMAKE_C_COMPILER}'.")
+if(WITH_DLT_QNX_SYSTEM AND NOT CMAKE_SYSTEM_NAME STREQUAL "QNX")
+    message(FATAL_ERROR "Can only compile for QNX with a QNX compiler")
 endif()
 
 if(WITH_DLT_FILE_LOGGING_SYSLOG_FALLBACK)
@@ -261,8 +261,21 @@ add_compile_options(
     -Wcast-qual
     -Wno-system-headers
     -Wno-unused-function
-    -Wno-stringop-truncation
 )
+if(NOT CMAKE_SYSTEM_NAME STREQUAL "QNX")
+    add_compile_options(-Wno-stringop-truncation)
+else()
+    add_compile_options(
+        -Wno-gnu-folding-constant
+        -Wno-strict-prototypes
+        -Wno-sign-conversion
+        -Wno-implicit-int-conversion
+        -Wno-gnu-zero-variadic-macro-arguments
+        -Wno-shorten-64-to-32
+        -Wno-unused-variable
+        -Wno-static-in-inline
+    )
+endif()
 
 if(WITH_DOC STREQUAL "OFF")
     set(PACKAGE_DOC "#")
