diff --color -U 6 -N -r before/src/common.h after/src/common.h
--- before/src/common.h	2026-04-08 21:54:22.878463825 -0400
+++ after/src/common.h	2026-04-08 21:54:47.674664793 -0400
@@ -63,13 +63,13 @@
 
 #define arraysize(array) (sizeof(ArraySizeHelper(array)))
 
 #if defined(_FREEBSD)
 #include <sys/endian.h>
 #endif
-#if !defined(__APPLE__) && !defined(_WIN32) && !defined(_FREEBSD)
+#if !defined(__APPLE__) && !defined(_WIN32) && !defined(_FREEBSD) && !defined(__QNX__)
 #include <endian.h>
 #if BYTE_ORDER == __BIG_ENDIAN
 #define IS_BIG_ENDIAN
 #endif
 #endif
 
