diff --git a/tensorflow/lite/kernels/CMakeLists.txt b/tensorflow/lite/kernels/CMakeLists.txt
index b9872323ac6..690695cbdfe 100644
--- a/tensorflow/lite/kernels/CMakeLists.txt
+++ b/tensorflow/lite/kernels/CMakeLists.txt
@@ -89,13 +89,13 @@ set(TEST_FRAMEWORK_SRC
   ${DELEGATE_PROVIDERS}
   ${TFLITE_SOURCE_DIR}/tools/optimize/model_utils.cc
   ${TF_SOURCE_DIR}/compiler/mlir/lite/tools/optimize/operator_property.cc
+  ${TF_SOURCE_DIR}/compiler/mlir/lite/tools/versioning/op_version.cc
   ${TF_SOURCE_DIR}/compiler/mlir/tools/optimize/quantization_utils.cc
   ${TF_SOURCE_DIR}/compiler/mlir/lite/kernels/internal/runtime_shape.cc
   ${TFLITE_SOURCE_DIR}/tools/optimize/quantization_utils.cc
   ${TFLITE_SOURCE_DIR}/tools/tool_params.cc
-  ${TFLITE_SOURCE_DIR}/tools/versioning/op_version.cc
   ${XLA_SOURCE_DIR}/xla/tsl/platform/default/env_time.cc
-  ${XLA_SOURCE_DIR}/xla/tsl/platform/default/logging.cc
+  ${XLA_SOURCE_DIR}/xla/tsl/platform/logging.cc
   internal/test_util.cc
   acceleration_test_util.cc
   acceleration_test_util_internal.cc
