diff --git a/runtime/lua/vim/fs.lua b/runtime/lua/vim/fs.lua
index d50ce39eac..77e32b9bf2 100644
--- a/runtime/lua/vim/fs.lua
+++ b/runtime/lua/vim/fs.lua
@@ -194,7 +194,18 @@ function M.dir(path, opts)
       if not fs then
         return
       end
-      return uv.fs_scandir_next(fs)
+
+      local name, etype = uv.fs_scandir_next(fs)
+      if not name then 
+        return
+      end
+
+      if etype == nil then
+        local stat = vim.uv.fs_lstat(M.joinpath(path, name))
+        etype = stat and stat.type or 'unknown'
+      end
+
+      return name, etype
     end
   end

