diff --git a/src/google/protobuf/compiler/command_line_interface.cc b/src/google/protobuf/compiler/command_line_interface.cc
index 03b970640..c62fab433 100644
--- a/src/google/protobuf/compiler/command_line_interface.cc
+++ b/src/google/protobuf/compiler/command_line_interface.cc
@@ -45,6 +45,8 @@
 #include <mach-o/dyld.h>
 #elif defined(__FreeBSD__)
 #include <sys/sysctl.h>
+#elif defined(__QNX__)
+#include <process.h>
 #endif
 
 #include "absl/algorithm/container.h"
@@ -222,6 +224,18 @@
   if (sysctl(mib, 4, &buffer, &len, nullptr, 0) != 0) {
     len = 0;
   }
+#elif defined(__QNX__)
+  char buffer[PATH_MAX];
+  int len = 0;
+  if (_cmdname(buffer) != nullptr) {
+    len = strlen(buffer);
+  }
 #else
   char buffer[PATH_MAX];
   int len = readlink("/proc/self/exe", buffer, PATH_MAX);
