--- a/CMakeLists.txt	2026-09-01 15:12:50.168558843 +0530
+++ b/CMakeLists.txt	2026-09-01 15:13:51.000000000 +0530
@@ -341,36 +341,50 @@
 endif(SNAPPY_BUILD_TESTS OR SNAPPY_BUILD_BENCHMARKS)
 
 if(SNAPPY_BUILD_TESTS)
+
   enable_testing()
 
-  # Prevent overriding the parent project's compiler/linker settings on Windows.
-  set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
-  set(install_gtest OFF)
-  set(install_gmock OFF)
-  set(build_gmock ON)
-
-  # This project is tested using GoogleTest.
-  add_subdirectory("third_party/googletest" EXCLUDE_FROM_ALL)
-
-  # GoogleTest triggers a missing field initializers warning.
-  if(SNAPPY_HAVE_NO_MISSING_FIELD_INITIALIZERS)
-    set_property(TARGET gtest
-        APPEND PROPERTY COMPILE_OPTIONS -Wno-missing-field-initializers)
-    set_property(TARGET gmock
-        APPEND PROPERTY COMPILE_OPTIONS -Wno-missing-field-initializers)
-  endif(SNAPPY_HAVE_NO_MISSING_FIELD_INITIALIZERS)
-
-  if(SNAPPY_HAVE_NO_IMPLICIT_INT_FLOAT_CONVERSION)
-    set_property(TARGET gtest
-        APPEND PROPERTY COMPILE_OPTIONS -Wno-implicit-int-float-conversion)
-  endif(SNAPPY_HAVE_NO_IMPLICIT_INT_FLOAT_CONVERSION)
+  find_library(GTEST_LIBRARY
+    NAMES gtest
+  )
+
+  find_library(GMOCK_LIBRARY
+    NAMES gmock
+  )
+
+  find_library(GMOCK_MAIN_LIBRARY
+    NAMES gmock_main
+  )
+
+  if(NOT GTEST_LIBRARY)
+    message(FATAL_ERROR "libgtest.so not found")
+  endif()
+
+  if(NOT GMOCK_LIBRARY)
+    message(FATAL_ERROR "libgmock.so not found")
+  endif()
+
+  if(NOT GMOCK_MAIN_LIBRARY)
+    message(FATAL_ERROR "libgmock_main.so not found")
+  endif()
+
+  message(STATUS "Using GTest: ${GTEST_LIBRARY}")
+  message(STATUS "Using GMock: ${GMOCK_LIBRARY}")
+  message(STATUS "Using GMock Main: ${GMOCK_MAIN_LIBRARY}")
 
   add_executable(snappy_unittest "")
+
   target_sources(snappy_unittest
     PRIVATE
       "snappy_unittest.cc"
   )
-  target_link_libraries(snappy_unittest snappy_test_support gmock_main gtest)
+
+  target_link_libraries(snappy_unittest
+    snappy_test_support
+    ${GMOCK_MAIN_LIBRARY}
+    ${GMOCK_LIBRARY}
+    ${GTEST_LIBRARY}
+  )
 
   add_test(
     NAME snappy_unittest
