--- gnome-keyring-48.0/egg/egg-unix-credentials.c
+++ a/egg/egg-unix-credentials.c
@@ -138,7 +138,7 @@
 		cred = (struct cmsgcred *) CMSG_DATA (&cmsg.hdr);
 		*pid = cred->cmcred_pid;
 		*uid = cred->cmcred_euid;
-#elif defined(LOCAL_CREDS)
+#elif defined(LOCAL_CREDS) && !defined(__QNX__)
 		cred = (struct sockcred *) CMSG_DATA (&cmsg.hdr);
 		*pid = 0;
 		*uid = cred->sc_euid;
@@ -230,7 +230,7 @@
 #if defined(LOCAL_CREDS) && !defined(HAVE_CMSGCRED)
 	int val = 1;
 	if (setsockopt (sock, 0, LOCAL_CREDS, &val, sizeof (val)) < 0) {
-		fprintf (stderr, "unable to set LOCAL_CREDS socket option on fd %d\n", fd);
+		fprintf (stderr, "unable to set LOCAL_CREDS socket option on fd %d\n", sock);
 		retval = -1;
 	}
 #endif
--- gnome-keyring-48.0/meson.build
+++ a/meson.build
@@ -57,13 +57,7 @@
 endif
 
 # Check for some types
-have_socklen_t = cc.has_type('socklen_t',
-  prefix: '\n'.join([
-    '#include <sys/types.h>',
-    '#include <sys/socket.h>',
-    '#include <netdb.h>',
-  ]),
-)
+have_socklen_t = true
 have_struct_cmsgcred = cc.has_type('struct cmsgcred',
   prefix: '\n'.join([
     '#include <sys/types.h>',
