diff --git a/src/meson.build b/src/meson.build
index f35c321..d14f7c6 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -415,7 +415,8 @@ else
             name: 'Pistache',
             description: 'An elegant C++ REST framework',
             url: 'https://pistacheio.github.io/pistache/',
-            version: '@0@-git@1@'.format(version_str, version_git_date),
+            # abuild does not like the version string with git date
+            version: '@0@'.format(version_str),
             filebase: 'libpistache',
             extra_cflags: public_args
     )
