diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0e976d0c1..b70d139d8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -89,7 +89,7 @@ else()
     find_package(Threads REQUIRED)
 endif()
 
-if(APPLE)
+if(APPLE OR QNX)
     set(OS_LIBS c Threads::Threads)
 elseif(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
     set(OS_LIBS thr execinfo)
diff --git a/tests/unit/s2n_rfc5952_test.c b/tests/unit/s2n_rfc5952_test.c
index 47c8f333a..8b957aebf 100644
--- a/tests/unit/s2n_rfc5952_test.c
+++ b/tests/unit/s2n_rfc5952_test.c
@@ -17,7 +17,7 @@
 
 #include "s2n_test.h"
 
-#if defined(__FreeBSD__) || defined(__OpenBSD__)
+#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__QNX__)
     #include <netinet/in.h>
     #include <sys/socket.h>
     #include <sys/types.h>
diff --git a/tls/s2n_ktls_io.c b/tls/s2n_ktls_io.c
index ad1d984c6..5b1726b54 100644
--- a/tls/s2n_ktls_io.c
+++ b/tls/s2n_ktls_io.c
@@ -211,7 +211,12 @@ S2N_RESULT s2n_ktls_sendmsg(void *io_context, uint8_t record_type, const struct
         .msg_iovlen = msg_iovlen,
     };
 
+#if defined(__QNX__) && (__QNX__ < 800)
+    char control_data[S2N_KTLS_CONTROL_BUFFER_SIZE];
+    memset(control_data, 0, S2N_KTLS_CONTROL_BUFFER_SIZE);
+#else
     char control_data[S2N_KTLS_CONTROL_BUFFER_SIZE] = { 0 };
+#endif
     RESULT_GUARD(s2n_ktls_set_control_data(&msg, control_data, sizeof(control_data),
             S2N_TLS_SET_RECORD_TYPE, record_type));
 
@@ -254,7 +259,12 @@ S2N_RESULT s2n_ktls_recvmsg(void *io_context, uint8_t *record_type, uint8_t *buf
      * member of the msghdr with the length of the control message
      * buffer.
      */
+#if defined(__QNX__) && (__QNX__ < 800)
+    char control_data[S2N_KTLS_CONTROL_BUFFER_SIZE];
+    memset(control_data, 0, S2N_KTLS_CONTROL_BUFFER_SIZE);
+#else
     char control_data[S2N_KTLS_CONTROL_BUFFER_SIZE] = { 0 };
+#endif
     msg.msg_controllen = sizeof(control_data);
     msg.msg_control = control_data;
 
