diff --git a/c++/CMakeLists.txt b/c++/CMakeLists.txt
index 9d7157f6..3b069d7b 100644
--- a/c++/CMakeLists.txt
+++ b/c++/CMakeLists.txt
@@ -160,7 +160,7 @@ else()
     message(SEND_ERROR "Cap'n Proto requires compiler-specific extensions (e.g., -std=gnu++14). Please leave CMAKE_CXX_EXTENSIONS undefined or ON.")
   endif()
 
-  if (NOT ANDROID)
+  if (NOT ANDROID AND NOT QNX)
     add_compile_options(-pthread)
   endif()
 endif()
diff --git a/c++/src/kj/CMakeLists.txt b/c++/src/kj/CMakeLists.txt
index ced85ca2..0632ff6b 100644
--- a/c++/src/kj/CMakeLists.txt
+++ b/c++/src/kj/CMakeLists.txt
@@ -78,8 +78,10 @@ add_library(kj ${kj_sources})
 add_library(CapnProto::kj ALIAS kj)
 target_compile_features(kj PUBLIC cxx_std_17)
 
-if(UNIX AND NOT ANDROID)
+if(UNIX AND NOT (ANDROID OR QNX))
   target_link_libraries(kj PUBLIC pthread)
+elseif(QNX)
+  target_link_libraries(kj PUBLIC socket)
 endif()
 
 include(CheckCXXSourceCompiles)
diff --git a/c++/src/kj/async-io-test.c++ b/c++/src/kj/async-io-test.c++
index bef5634b..e2113b83 100644
--- a/c++/src/kj/async-io-test.c++
+++ b/c++/src/kj/async-io-test.c++
@@ -2921,6 +2921,7 @@ KJ_TEST("Userspace TwoWayPipe whenWriteDisconnected()") {
 
 #if !_WIN32  // We don't currently support detecting disconnect with IOCP.
 #if !__CYGWIN__  // TODO(someday): Figure out why whenWriteDisconnected() doesn't work on Cygwin.
+#if !__QNX__ // POLLHUP is not triggered for disconnected pipes in QNX
 
 KJ_TEST("OS OneWayPipe whenWriteDisconnected()") {
   auto io = setupAsyncIo();
@@ -2981,6 +2982,7 @@ KJ_TEST("import socket FD that's already broken") {
   KJ_EXPECT(buffer == "foo"_kj);
 }
 
+#endif  // !__QNX__
 #endif  // !__CYGWIN__
 #endif  // !_WIN32
 
diff --git a/c++/src/kj/async-io-unix.c++ b/c++/src/kj/async-io-unix.c++
index 7e1c85fc..5ac11db1 100644
--- a/c++/src/kj/async-io-unix.c++
+++ b/c++/src/kj/async-io-unix.c++
@@ -1357,8 +1357,8 @@ public:
               ownFd.get(), IPPROTO_TCP, TCP_NODELAY, (char*)&one, sizeof(one))) {
           case EOPNOTSUPP:
           case ENOPROTOOPT: // (returned for AF_UNIX in cygwin)
-#if __FreeBSD__
-          case EINVAL: // (returned for AF_UNIX in FreeBSD)
+#if __FreeBSD__ || __QNX__
+          case EINVAL: // (returned for AF_UNIX in FreeBSD and QNX)
 #endif
             break;
           default:
diff --git a/c++/src/kj/async-unix-test.c++ b/c++/src/kj/async-unix-test.c++
index 64190c43..9da4c71b 100644
--- a/c++/src/kj/async-unix-test.c++
+++ b/c++/src/kj/async-unix-test.c++
@@ -573,6 +573,17 @@ TEST(AsyncUnixTest, WriteObserver) {
     KJ_NONBLOCKING_SYSCALL(n = write(outfd, "foo", 3));
   } while (n >= 0);
 
+#ifdef __QNX__
+  // On QNX, there is a user-customizable "atomic write size" for pipes
+  // writes below this size either succeed or fail as a whole; there
+  // is no "partial writes" when atomic. As a result, to truly fill
+  // the buffer, we need to add as many 1 bytes as possible after we
+  // are done with filling the buffer the normal way.
+  do {
+    KJ_NONBLOCKING_SYSCALL(n = write(outfd, "a", 1));
+  } while (n >= 0);
+#endif
+
   bool writable = false;
   auto promise = observer.whenBecomesWritable()
       .then([&]() { writable = true; }).eagerlyEvaluate(nullptr);
@@ -599,7 +610,7 @@ TEST(AsyncUnixTest, WriteObserver) {
   EXPECT_TRUE(writable);
 }
 
-#if !__APPLE__ && !(KJ_USE_KQUEUE && !defined(EVFILT_EXCEPT))
+#if !__APPLE__ && !__QNX__ && !(KJ_USE_KQUEUE && !defined(EVFILT_EXCEPT))
 // Disabled on macOS due to https://github.com/capnproto/capnproto/issues/374.
 // Disabled on kqueue systems that lack EVFILT_EXCEPT because it doesn't work there.
 TEST(AsyncUnixTest, UrgentObserver) {
@@ -978,6 +989,8 @@ KJ_TEST("UnixEventPort whenWriteDisconnected()") {
   hupPromise.wait(waitScope);
 }
 
+#if !__QNX__
+
 KJ_TEST("UnixEventPort FdObserver(..., flags=0)::whenWriteDisconnected()") {
   // Verifies that given `0' as a `flags' argument,
   // FdObserver still observes whenWriteDisconnected().
@@ -1005,7 +1018,8 @@ KJ_TEST("UnixEventPort FdObserver(..., flags=0)::whenWriteDisconnected()") {
   hupPromise.wait(waitScope);
 }
 
-#endif
+#endif // !__QNX__
+#endif // !__CYGWIN__
 
 KJ_TEST("UnixEventPort poll for signals") {
   captureSignals();
diff --git a/c++/src/kj/async.c++ b/c++/src/kj/async.c++
index 496a72cc..dfc26220 100644
--- a/c++/src/kj/async.c++
+++ b/c++/src/kj/async.c++
@@ -56,7 +56,16 @@
 #else
 
 #if KJ_USE_FIBERS
+#ifdef __QNX__
+#include <libucontext/libucontext.h>
+#define getcontext libucontext_getcontext
+#define makecontext libucontext_makecontext
+#define setcontext libucontext_setcontext
+#define swapcontext libucontext_swapcontext
+#define ucontext_t libucontext_ucontext_t
+#else
 #include <ucontext.h>
+#endif
 #include <setjmp.h>    // for fibers
 #endif
 
diff --git a/c++/src/kj/exception.c++ b/c++/src/kj/exception.c++
index 29405f64..709e5ed9 100644
--- a/c++/src/kj/exception.c++
+++ b/c++/src/kj/exception.c++
@@ -84,6 +84,10 @@
 #include <ucontext.h>
 #endif
 
+#if __QNX__
+#include <sys/nto_version.h>
+#endif
+
 #if KJ_HAS_LIBDL
 #include "dlfcn.h"
 #endif
@@ -585,9 +589,10 @@ void printStackTraceOnCrash() {
 #endif
 }
 
-#elif _WIN32
+#elif _WIN32 || __QNX__ < 800
 // Windows, but KJ_USE_WIN32_DBGHELP is not enabled. We can't print useful stack traces, so don't
 // try to catch SEH nor ctrl+C.
+// For QNX before 8.0.0, no signalstack support is present.
 
 void printStackTraceOnCrash() {
 #if !KJ_NO_EXCEPTIONS
@@ -1241,7 +1246,7 @@ struct FakeEhGlobals {
 
 // LLVM's libstdc++ doesn't declare __cxa_get_globals in its cxxabi.h. GNU does. Because it is
 // extern "C", the compiler wills get upset if we re-declare it even in a different namespace.
-#if _LIBCPPABI_VERSION
+#if _LIBCPPABI_VERSION || __QNX__
 extern "C" void* __cxa_get_globals();
 #else
 using abi::__cxa_get_globals;
diff --git a/c++/src/kj/filesystem-disk-test.c++ b/c++/src/kj/filesystem-disk-test.c++
index bb7614c8..5dee3912 100644
--- a/c++/src/kj/filesystem-disk-test.c++
+++ b/c++/src/kj/filesystem-disk-test.c++
@@ -235,7 +235,7 @@ public:
 
   Own<Directory> get() {
     int fd;
-    KJ_SYSCALL(fd = open(filename.cStr(), O_RDONLY));
+    KJ_SYSCALL(fd = open(filename.cStr(), O_RDONLY | O_DIRECTORY));
     return newDiskDirectory(AutoCloseFd(fd));
   }
 
@@ -346,6 +346,7 @@ KJ_TEST("DiskFile") {
     KJ_EXPECT(kj::str(writableMapping->get().slice(0, 6).asChars()) == "fDobaz");
     KJ_EXPECT(kj::str(privateMapping.slice(0, 6).asChars()) == "Foobaz");
 
+#if !__QNX__ // Usage of write() with writable mmap() is unsupported in QNX
     file->write(0, StringPtr("qux").asBytes());
     KJ_EXPECT(kj::str(mapping.slice(0, 6).asChars()) == "quxbaz");
     KJ_EXPECT(kj::str(writableMapping->get().slice(0, 6).asChars()) == "quxbaz");
@@ -353,6 +354,7 @@ KJ_TEST("DiskFile") {
 
     file->write(12, StringPtr("corge").asBytes());
     KJ_EXPECT(kj::str(mapping.slice(12, 17).asChars()) == "corge");
+#endif
 
 #if !_WIN32 && !__CYGWIN__  // Windows doesn't allow the file size to change while mapped.
     // Can shrink.
@@ -370,9 +372,17 @@ KJ_TEST("DiskFile") {
 
   file->truncate(6);
 
+#if !__QNX__
   KJ_EXPECT(file->readAllText() == "quxbaz");
+#else
+  KJ_EXPECT(file->readAllText() == "fDobaz");
+#endif
   file->zero(3, 3);
+#if !__QNX__
   KJ_EXPECT(file->readAllText() == StringPtr("qux\0\0\0", 6));
+#else
+  KJ_EXPECT(file->readAllText() == StringPtr("fDo\0\0\0", 6));
+#endif
 }
 
 KJ_TEST("DiskFile::copy()") {
@@ -765,7 +775,9 @@ KJ_TEST("DiskDirectory replaceSubdir()") {
     KJ_EXPECT(!dir->exists(Path({"foo", "bar"})));
 
     replacer->commit();
+#if !__QNX__ // QNX invalidates dirfd after rename
     KJ_EXPECT(replacer->get().openFile(Path("bar"))->readAllText() == "original");
+#endif
     KJ_EXPECT(dir->openFile(Path({"foo", "bar"}))->readAllText() == "original");
   }
 
@@ -997,7 +1009,7 @@ KJ_TEST("DiskFilesystem::computeCurrentPath") {
     "some_path_longer_than_256_bytes"
   }), WriteMode::CREATE | WriteMode::CREATE_PARENT);
 
-  auto origDir = open(".", O_RDONLY);
+  auto origDir = open(".", O_RDONLY | O_DIRECTORY);
   KJ_SYSCALL(fchdir(KJ_ASSERT_NONNULL(subdir->getFd())));
   KJ_DEFER(KJ_SYSCALL(fchdir(origDir)));
 
diff --git a/c++/src/kj/filesystem-disk-unix.c++ b/c++/src/kj/filesystem-disk-unix.c++
index 3b42ed87..fff67b12 100644
--- a/c++/src/kj/filesystem-disk-unix.c++
+++ b/c++/src/kj/filesystem-disk-unix.c++
@@ -416,8 +416,8 @@ public:
 
     static const byte ZEROS[4096] = { 0 };
 
-#if __APPLE__ || __CYGWIN__ || (defined(__ANDROID__) && __ANDROID_API__ < 24)
-    // Mac & Cygwin & Android API levels 23 and lower doesn't have pwritev().
+#if __APPLE__ || __CYGWIN__ || __QNX__ || (defined(__ANDROID__) && __ANDROID_API__ < 24)
+    // Mac & Cygwin & QNX & Android API levels 23 and lower doesn't have pwritev().
     while (size > sizeof(ZEROS)) {
       write(offset, ZEROS);
       size -= sizeof(ZEROS);
@@ -672,8 +672,10 @@ public:
   template <typename Func>
   auto list(bool needTypes, Func&& func) const
       -> Array<Decay<decltype(func(instance<StringPtr>(), instance<FsNode::Type>()))>> {
+#if !__QNX__
     // Seek to start of directory.
     KJ_SYSCALL(lseek(fd, 0, SEEK_SET));
+#endif
 
     // Unfortunately, fdopendir() takes ownership of the file descriptor. Therefore we need to
     // make a duplicate.
diff --git a/c++/src/kj/miniposix.h b/c++/src/kj/miniposix.h
index e9ae848d..d79b6f3f 100644
--- a/c++/src/kj/miniposix.h
+++ b/c++/src/kj/miniposix.h
@@ -132,6 +132,10 @@ static constexpr inline size_t iovMax() {
 static constexpr inline size_t iovMax() {
   return UIO_MAX_IOV;
 }
+#elif __QNX__
+static constexpr inline size_t iovMax() {
+  return 1024;
+}
 #else
 #error "Please determine the appropriate constant for IOV_MAX on your system."
 #endif
