--- a/Source/WTF/wtf/glib/FileSystemGlib.cpp
+++ b/Source/WTF/wtf/glib/FileSystemGlib.cpp
@@ -76,6 +76,18 @@
 {
     return { };
 }
+#elif OS(QNX)
+#include <process.h>
+
+CString currentExecutablePath()
+{
+    static char readBuffer[PATH_MAX];
+    if (!_cmdname(readBuffer))
+        return { };
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN // QNX port
+    return CString( std::span{ readBuffer, strlen(readBuffer) } );
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+}
 #elif OS(UNIX)
 CString currentExecutablePath()
 {
