--- a/Source/cmake/WebKitCommon.cmake
+++ b/Source/cmake/WebKitCommon.cmake
@@ -99,6 +99,12 @@
     else ()
         string(TOLOWER ${CMAKE_SYSTEM_PROCESSOR} LOWERCASE_CMAKE_SYSTEM_PROCESSOR)
     endif ()
+    
+    # QNX reports "unknown" as CMAKE_SYSTEM_PROCESSOR, use uname instead
+    if (CMAKE_SYSTEM_NAME MATCHES "QNX" AND LOWERCASE_CMAKE_SYSTEM_PROCESSOR STREQUAL "unknown")
+        execute_process(COMMAND uname -m OUTPUT_VARIABLE QNX_UNAME_PROCESSOR OUTPUT_STRIP_TRAILING_WHITESPACE)
+        string(TOLOWER ${QNX_UNAME_PROCESSOR} LOWERCASE_CMAKE_SYSTEM_PROCESSOR)
+    endif ()
     if (LOWERCASE_CMAKE_SYSTEM_PROCESSOR MATCHES "(^aarch64|^arm64|^cortex-?[am][2-7][2-8])")
          if (FORCE_32BIT)
             set(WTF_CPU_ARM 1)
@@ -111,7 +117,7 @@
         set(WTF_CPU_MIPS64 1)
     elseif (LOWERCASE_CMAKE_SYSTEM_PROCESSOR MATCHES "^mips")
         set(WTF_CPU_MIPS 1)
-    elseif (LOWERCASE_CMAKE_SYSTEM_PROCESSOR MATCHES "(x64|x86_64|amd64)")
+    elseif (LOWERCASE_CMAKE_SYSTEM_PROCESSOR MATCHES "(x64|x86_64|amd64|x86pc)")
         # FORCE_32BIT is set in the build script when --32-bit is passed
         # on a Linux/intel 64bit host. This allows us to produce 32bit
         # binaries without setting the build up as a crosscompilation,
--- a/Source/WebKit/NetworkProcess/soup/NetworkProcessSoup.cpp	2026-02-06 14:50:34.000000000 +0000
+++ b/Source/WebKit/NetworkProcess/soup/NetworkProcessSoup.cpp	2026-02-06 14:51:26.000000000 +0000
@@ -125,6 +125,7 @@
void NetworkProcess::platformInitializeNetworkProcess(const NetworkProcessCreationParameters& parameters)
{
+#ifndef __QNX__
     GRefPtr<GResolver> cachedResolver = adoptGRef(webkitCachedResolverNew(adoptGRef(g_resolver_get_default())));
     g_resolver_set_default(cachedResolver.get());
@@ -132,7 +133,7 @@
         auto overridingResolver = adoptGRef(webkitOverridingResolverNew(cachedResolver.get(), parameters.localhostAliasesForTesting));
         g_resolver_set_default(overridingResolver.get());
     }
-
+#endif
     m_cacheOptions = { NetworkCache::CacheOption::RegisterNotify };
     // FIXME: NetworkProcess probably does not have session at this point.
     forEachNetworkSession([&](NetworkSession& session) {
