diff --color -U 6 -r before/tests/test_invalid_binary.py after/tests/test_invalid_binary.py
--- before/tests/test_invalid_binary.py	2026-03-31 16:08:59.507591974 -0400
+++ after/tests/test_invalid_binary.py	2026-03-31 16:12:01.141101903 -0400
@@ -61,11 +61,12 @@
                 # This is what should happen
                 pass
             else:
                 # Re-raise the original error to fail the test
                 raise
         finally:
-            os.unlink(fullpath)
-            os.rmdir(dirpath)
+            pass
+            #os.unlink(fullpath)
+            #os.rmdir(dirpath)
 
 if __name__ == '__main__':
     unittest.main()
